Core/tests/WholeProjectRefactorer-ApplyRefactoringForVariablesContainer.cpp

```diff
@@ -0,0 +1,553 @@
+/*
+ * GDevelop Core
+ * Copyright 2008-2016 Florian Rival (Florian.Rival@gmail.com). All rights
+ * reserved. This project is released under the MIT License.
+ */
+/**
+ * @file Tests covering project refactoring
+ */
+#include <algorithm>
+#include <stdexcept>
+
+#include "DummyPlatform.h"
+#include "GDCore/Events/Builtin/ForEachChildVariableEvent.h"
+#include "GDCore/Events/Builtin/LinkEvent.h"
+#include "GDCore/Events/Builtin/RepeatEvent.h"
+#include "GDCore/Events/Builtin/StandardEvent.h"
+#include "GDCore/Events/Event.h"
+#include "GDCore/Extensions/Metadata/MetadataProvider.h"
+#include "GDCore/Extensions/Metadata/ParameterMetadataTools.h"
+#include "GDCore/Extensions/Platform.h"
+#include "GDCore/Extensions/PlatformExtension.h"
+#include "GDCore/IDE/WholeProjectRefactorer.h"
+#include "GDCore/Project/Behavior.h"
+#include "GDCore/Project/EventsFunctionsExtension.h"
+#include "GDCore/Project/ExternalEvents.h"
+#include "GDCore/Project/ExternalLayout.h"
+#include "GDCore/Project/Layout.h"
+#include "GDCore/Project/Object.h"
+#include "GDCore/Project/Project.h"
+#include "GDCore/Project/Variable.h"
+#include "GDCore/Serialization/Serializer.h"
+#include "catch.hpp"
+
+TEST_CASE("WholeProjectRefactorer::ApplyRefactoringForVariablesContainer",
+          "[common]") {
+  SECTION("Variable renamed (project, layout, object)") {
+    gd::Project project;
+    gd::Platform platform;
+    SetupProjectWithDummyPlatform(project, platform);
+    auto &layout1 = project.InsertNewLayout("Layout1", 0);
+    gd::StandardEvent &event =
+        dynamic_cast<gd::StandardEvent &>(layout1.GetEvents().InsertNewEvent(
+            project, "BuiltinCommonInstructions::Standard"));
+    gd::ForEachChildVariableEvent &forEachChildVariableEvent =
+        dynamic_cast<gd::ForEachChildVariableEvent &>(
+            layout1.GetEvents().InsertNewEvent(
+                project, "BuiltinCommonInstructions::ForEachChildVariable"));
+    gd::RepeatEvent &repeatEvent =
+        dynamic_cast<gd::RepeatEvent &>(layout1.GetEvents().InsertNewEvent(
+            project, "BuiltinCommonInstructions::Repeat"));
+
+    // Declare global variables.
+    project.GetVariables().InsertNew("MyGlobalVariable", 0).SetValue(123);
+    project.GetVariables().InsertNew("MyGlobalVariable2", 0).SetValue(123);
+    project.GetVariables().InsertNew("SharedVariableName", 0).SetValue(123);
+    project.GetVariables()
+        .InsertNew("MyGlobalStructureVariable", 0)
+        .GetChild("MyChild")
+        .SetValue(123);
+    project.GetVariables()
+        .InsertNew("MyGlobalStructureVariable2", 0)
+        .GetChild("MyChild")
+        .SetValue(123);
+
+    // Declare variables in the scene.
+    layout1.GetVariables().InsertNew("MySceneVariable", 0).SetValue(123);
+    layout1.GetVariables().InsertNew("MySceneVariable2", 0).SetValue(123);
+    layout1.GetVariables().InsertNew("SharedVariableName", 0).SetValue(123);
```

Comment:
This is about "global variable vs scene variable", is there cases for other name collisions?
- object variable > scene child variable > global child variable